### ISSUE TITLE
Python 3.13 can't compile with armv5 target

### ISSUE THREAD

**Issue by rossburton** [python/cpython#125444]
# Bug report

### Bug description:

If we build cpython 3.13 for Linux/ARMv5 using `-march=armv5te` and then run inside a qemu with the `versatilepb` machine (which uses a ARM926EJ-S processor) then python crashes with "Illegal Instruction".  Specifically, `mrc` is used in a few places to get the thread ID, but this not possible in cores lower than ARMv6K.

Users of `mrc`:
- https://github.com/python/cpython/blob/main/Include/object.h#L195
- https://github.com/python/cpython/blob/main/Include/internal/mimalloc/mimalloc/prim.h#L172
- https://github.com/python/cpython/blob/main/Include/internal/mimalloc/mimalloc/prim.h#L199

I suspect the best way of implemented the required check would be:

```
#if defined(__arm__) && __ARM_ARCH >= 6 && !defined(__ARM_ARCH_6__)
```

That is, if Arm and if the architecture level is 6 or great and if this is not bare v6 (so v6K/v6Z are included).

Or a blunter but simpler hammer on the grounds that v5/v6 is pretty niche these days could be:

```
#if defined(__arm__) && __ARM_ARCH >= 7
```

### CPython versions tested on:

3.13

### Operating systems tested on:

Linux

### Linked PRs
* gh-125574
* gh-125595

**Comment by rossburton** [comment]
Paging @diegorusso.

**Comment by diegorusso** [comment]
On Arm v5 it is not possible to get the thread ID via register hence the illegal instruction. The c13 register started to provide thread ID [since the rev1 (r1p0) release of the ARM1136JF-S processor](https://developer.arm.com/documentation/ddi0211/k/system-control-coprocessor/system-control-coprocessor-register-descriptions/c13--thread-and-process-id-registers?lang=en) which is an Arm v6K architecture varian. Other variants of Arm v6 (T2, Z and base ) don’t provide the thread ID via c13.
Said that, for the sake of simplicity here we could group v5 and v6 together and consider that instructions for v7 only (as @rossburton said).
This is also similarly done here: https://github.com/python/cpython/blob/main/Include/internal/mimalloc/mimalloc/atomic.h#L351

**Comment by colesbury** [comment]
Does this affect the default build of Python 3.13 or just the free threading build? `_Py_ThreadId()` is only compiled in the free threading build.

Here's glibc's check, for what it's worth: https://github.com/bminor/glibc/blob/9d4b4515a88c5d0bbfc7809374f322c507c2d779/sysdeps/arm/sysdep.h#L59-L60

And their fallback:
https://github.com/bminor/glibc/blob/9d4b4515a88c5d0bbfc7809374f322c507c2d779/sysdeps/arm/sysdep.h#L265-L276

**Comment by kanavin** [comment]
> Does this affect the default build of Python 3.13 or just the free threading build? `_Py_ThreadId()` is only compiled in the free threading build.

We only test the default build, and indeed the issue in that build is seen only with the usage of the instruction in mimalloc. I've disabled mimalloc on armv5 as a workaround.

**Comment by diegorusso** [comment]
We are currently testing a patch that draws a line between Arm v6 and Arm v7. I guess we can go with a more precise grouping like glibc does but I feel it is not that necessary for a couple of reasons:
* **code maintenance**: is it worth to have different code paths compared with the number of use cases where an Arm v6k/v6zk board needs to run Python with mimalloc/free threading? This differentiation will fade out in the coming years and by the time free threading will be enabled by default, the difference will be even more shallow.
* **testing**: how are we going to have proper testing? Even now we are relying on software emulation to test it because it's already hard to fetch different variants of Arm v5/v6 boards.

glibc has that differentiation because it was introduce organically whilst being developed over the years (that differentiation [was introduced 11 years ago and never changed it](https://github.com/bminor/glibc/commit/3377126b0a055e6addaa6bd69f51dd557d774b99)) and their main focus is back compatibility.
The backward compatibility that we have with mimalloc is not that deeper as the glibc one, and a division between Arm v6 and Arm v7 should suffice.

Anyway if anyone feels strong otherwise, we could push for a more precise grouping.

**Comment by diegorusso** [comment]
I have created the PR, please have a look.

**Comment by rossburton** [comment]
Looks good to me!

**Comment by diegorusso** [comment]
Changes have been merged both on main and 3.13 branches.
I will port the fix to mimalloc in the next days.
